build-tools/packages/build-cli/src/commands/generate/manifest.ts

```diff
@@ -0,0 +1,226 @@
+					buildNumber,
+					this.logger,
+				);
+				if (devVersion !== undefined) {
+					this.log(`Fetched dev version: ${devVersion}`);
+					await generateManifestFileForDevReleases(devVersion, this.logger);
+				}
+			} else if (buildNumber === undefined) {
+				this.log(
+					`No successful build found for ${flags.sourceBranch} branch in the last 24 hours`,
+				);
+			}
+		} catch (error: unknown) {
+			this.errorLog(`Error creating manifest file: ${error}`);
+		}
+	}
+}
+
+/**
+ * Fetches the first successful build number in the last 24 hours.
+ * @returns The build number if successful, otherwise undefined.
+ */
+async function getFirstSuccessfulBuild(
+	authHeader: string,
+	organization: string,
+	project: string,
+	sourceBranch: string,
+	log?: Logger,
+): Promise<string | undefined> {
+	const ADO_BASE_URL = `https://dev.azure.com/${organization}/${project}/_apis/build/builds?api-version=7.0`;
+	try {
+		const response = await fetch(ADO_BASE_URL, { headers: { Authorization: authHeader } });
+		const data = await response.json();
+
+		const twentyFourHoursAgo = new Date();
+		twentyFourHoursAgo.setHours(twentyFourHoursAgo.getHours() - 24);
+
+		const successfulBuilds = data.value.filter(
+			(build: IBuildDetails) =>
+				build.definition.name === "Build - client packages" &&
+				build.status === "completed" &&
+				build.result === "succeeded" &&
+				build.sourceBranch === `refs/heads/${sourceBranch}` &&
+				new Date(build.finishTime) >= twentyFourHoursAgo,
+		);
+
+		return successfulBuilds.length > 0
+			? (successfulBuilds[0].buildNumber as string)
+			: undefined;
+	} catch (error) {
+		log?.errorLog("Error fetching successful builds:", error);
+		return undefined;
+	}
+}
+
+/**
+ * Fetches the dev version number released in the build feed.
+ * @param buildNumber - The build number.
+ * @returns The dev version number if found, otherwise undefined.
+ */
+async function fetchDevVersionNumber(
+	authHeader: string,
+	organization: string,
+	project: string,
+	buildNumber: string,
+	log?: Logger,
+): Promise<string | undefined> {
+	const REGISTRY_URL = `https://pkgs.dev.azure.com/${organization}/${project}/_packaging/build/npm/registry/`;
+	try {
+		const response = await fetch(`${REGISTRY_URL}/${PACKAGE_NAME}`, {
+			headers: { Authorization: authHeader },
+		});
+		const data = await response.json();
+		const buildVersionKey = Object.keys(data.time).find((key) => key.includes(buildNumber));
+
+		if (buildVersionKey !== undefined) {
+			return buildVersionKey;
+		}
+
+		log?.log(`No version with build number ${buildNumber} found.`);
+		return undefined;
+	} catch (error: unknown) {
+		log?.errorLog("Error fetching dev version number:", error);
+		return undefined;
+	}
+}
+
+/**
+ * Generates a modified manifest file with the specified version number.
+ * @param version - The version number.
+ */
+async function generateManifestFileForDevReleases(
+	version: string,
+	log?: Logger,
+): Promise<string | undefined> {
```

Comment:
Separate this into two functions - one that generates the manifest as an object and a second one that writes to the files. Then you can test the manifest generation without writing the file.